### PR TITLE
Dependency version consolidation

### LINKED ISSUE #89: Different declared dependency versions between subprojects
# Issue
Due to the nature of multi project builds, dependencies will be declared multiple times. This can and does lead to different versions of dependencies being declared.

For this project, the following dependencies have multiple different version declared:
```
    slf4j-api: {'1.7.25', '1.7.26'}
        guava: {'11.0.2', '19.0'}
       hsqldb: {'2.5.0', '2.4.1'}
slf4j-log4j12: {'1.7.25', '1.7.25 '}
commons-lang3: {'3.8', '3.2'}
   log4j-core: {'2.11.0', '2.11.2'}
    log4j-api: {'2.11.0', '2.11.2'}
    oshi-core: {'3.13.2', '4.0.0'}
```

This can, but doesn't have to, lead to issues both at compile and runtime, as different dependency version might be incompatible, especially if they have a mismatch in their major version.

# Possible solutions

## `subprojects` declaration
Dependencies could be declared via a root `build.gradle` file, which then applies the depenencies to all subprojects. This probably isn't a good solution for this project, as there are a lot of dependencies that are only required for some of the subprojects, but not all of them.

## `extra properties`
Probably the more appropriate solution for this project is to declare the dependency versions as so called [extra properties](https://docs.gradle.org/current/userguide/writing_build_scripts.html#sec:extra_properties) in a root `build.gradle` file.

This means that the used version is only declared once for the entire build. The subprojects still declare their own dependencies, but use the extra properties for the version number.

#### Example

`build.gradle`
```groovy
ext {
    junit_version = "4.12"
    slf4j_api_version = "1.7.26"
}
```

`core/build.gradle`
```groovy
...
dependencies {
    compile group: "org.slf4j", name: "slf4j-api", version: slf4j_api_version
    testCompile group: "junit", name: "junit", version: junit_version
}
...
```

## Helpful script
I have written [this script](https://gist.github.com/schoenja/21f72983696566bd9bd9a77a1570aec4) to read out the used dependencies and version numbers for the build.

### PR DESCRIPTION
**Pull request by schoenja**
Moves (hopefully) all dependencies to a single version number.
This will fix #89 .

Possible improvements:
- Some dependencies share a version number (e.g. `log4j_*`), this could be merged into a single variable
- buildscript dependencies are not managed

**IMPORTANT**: I haven't been able to test whether these changes break anything!

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    lang_painless_version = "6.2.4"
+    log4j_api_version = "2.11.2" // alternative version: 2.11.0
+    log4j_core_version = "2.11.2" // alternative version: 2.11.0
+    log4j_slf4j_impl_version = "2.11.2"
+    mockito_core_version = "2.22.0"
+    mongo_java_driver_version = "3.5.0"
+    mysql_connector_java_version = "5.1.20"
+    natty_version = "0.13"
+    opencsv_version = "2.3"
+    oshi_core_version = "4.0.0" // alternative version: 3.13.2
```
Comment (vogti) [inline build.gradle]: Please use version 3.13.2 instead.

Comment (schoenja) [reply build.gradle]: Trying that out right now, but first testing what happens if the `dbms` subproject is downgraded from `4.0.0` to `3.13.2`

**Comment by vogti** [comment]
Is there a specific reasons for defining the variables in the ext-block of a root `build.gradle` file instead of adding them to the `gradle.properties` file?

**Comment by schoenja** [comment]
> Is there a specific reasons for defining the variables in the ext-block of a root `build.gradle` file instead of adding them to the `gradle.properties` file?

I just double checked: using the `gradle.properties` file is fine for this. I misunderstood part of the gradle documentation, which is why I had added the root `build.gradle`.

Side note: switching to `gradle.properties` means we are affected by this [IntelliJ inspection bug](https://youtrack.jetbrains.com/issue/IDEA-146200), which marks all properties as "unused".

**Comment by vogti** [comment]
Thank you very much @schoenja for this PR!

**Pull request merged**